llvm/test/Transforms/InstCombine/sqrt.ll

```diff
@@ -201,6 +201,25 @@ define <2 x float> @sqrt_exp_vec(<2 x float> %x) {
   ret <2 x float> %res
 }
 
+define dso_local fp128 @sqrt_exp_long_double(fp128 %x, fp128 %y) {
+; CHECK-LABEL: @sqrt_exp_long_double(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[MERGED_SQRT:%.*]] = fmul fast fp128 [[X:%.*]], 0xL00000000000000003FFE000000000000
+; CHECK-NEXT:    [[TMP0:%.*]] = call fast fp128 @llvm.exp.f128(fp128 [[MERGED_SQRT]])
+; CHECK-NEXT:    [[MERGED_SQRT1:%.*]] = fmul fast fp128 [[Y:%.*]], 0xL00000000000000003FFE000000000000
+; CHECK-NEXT:    [[TMP1:%.*]] = call fast fp128 @llvm.exp2.f128(fp128 [[MERGED_SQRT1]])
+; CHECK-NEXT:    [[ADD:%.*]] = fadd fast fp128 [[TMP0]], [[TMP1]]
+; CHECK-NEXT:    ret fp128 [[ADD]]
+;
+entry:
+  %0 = call fast fp128 @llvm.exp.f128(fp128 %x)
+  %1 = call fast fp128 @llvm.sqrt.f128(fp128 %0)
+  %2 = call fast fp128 @llvm.exp2.f128(fp128 %y)
+  %3 = call fast fp128 @llvm.sqrt.f128(fp128 %2)
+  %add = fadd fast fp128 %1, %3
```

Comment:
Test the different cases in different functions.

Use minimal fast math flags, this looks like it only requires reassoc. Add an additional test which shows the preservation behavior of other fast math flags.

Also test the same situation with equivalent library calls instead of intrinsics (and the mixed intrinsic + libcall case)